README.rst
dropafile
=========

Drop me a file, securely.

|build-status|_ | `documentation <http://dropafile.readthedocs.org/>`_ | `sources <https://github.com/ulif/dropafile>`_ | `issues <https://github.com/ulif/dropafile/issues>`_

.. |build-status| image:: https://travis-ci.org/ulif/dropafile.png?branch=master
.. _build-status: https://travis-ci.org/ulif/dropafile


`dropafile` provides an HTTPS-secured webapp where users can drop
files.

It is meant as a channel to deliver documents in a not-too-unsecure
manner. For instance as a quickly installable workaround if people are
not able or willing to use GnuPG or similar, although they have
sensible documents to send.

`dropafile` is written in Python (server parts) and uses the
`dropzonejs`_ JavaScript library (client parts). The builtin server is
based on Werkzeug_.

Install
-------

As a user, run::

  $ pip install dropafile

then, start the local server::

  $ dropafile
  Creating temporary self-signed SSL certificate...
  Done.
  Certificate in: /tmp/tmp1y2bgh/cert.pem
  Key in:         /tmp/tmp1y2bgh/cert.key
  Password is: H93rqnsrdEXD2ad3rQwdWqZ
   * Running on https://localhost:8443/ (Press CTRL+C to quit)

The server will provide SSL. Users can access `dropafile` sevice
pointing their browsers to the location given. The page is protected
by basic auth. Users will have to provide an arbitrary user name and
the password displayed on the commandline at startup (which changes
with restart).

The `--help` option will display all available options::

  $ dropfile --help
  usage: dropafile [-h] [--host HOST] [-p PORT] [-s PASSWORD]

  Start dropafile app.

  optional arguments:
    -h, --help            show this help message and exit
    --host HOST           Host we bind to. An IP address or DNS name.
                          `localhost` by default.
    -p PORT, --port PORT  Port we listen at. An integer. 8443 by default.
    -s PASSWORD, --secret PASSWORD
                          Password to access dropafile. If none is given we
                          generate one.

Whenever a user sends a file, the path is displayed on the
commandline.


Developer Install
-----------------

Developers should install a `virtualenv`_ first::

  $ virtualenv -p /usr/bin/python2.7 py27  # for Python2.7

See `tox.ini` for all Python versions supported.

Activate the virtualenv::

  $ source py27/bin/activate
  (py27) $

Now build the devel environment::

  (py27) $ python setup.py dev

You can run tests like this::

  (py27) $ py.test

Tests for all supported (and locally available) Python vesions can be
run by::

  (py27) $ pip install tox  # neccessary only once per virtualenv
  (py27) $ tox


.. _virtualenv: https://virtualenv.pypa.io/
.. _dropzonejs: http://www.dropzonejs.com/
.. _Werkzeug: http://werkzeug.pocoo.org/


CHANGES.rst
Changes
=======

0.1.2dev (unreleased)
---------------------

- No changes yet.


0.1.1 (2015-03-30)
------------------

- Turned former `dropafile` module into a Python package. This is to
  fix installation behvior where data files are installed in different
  locations depending on install tool. See `diceware bug #1
  <https://github.com/ulif/diceware/issues/1>`_.

  The problem was revealed by `conorsch
  <https://github.com/conorsch>`_ for the `diceware
  <https://github.com/ulif/diceware>`_ package.


0.1 (2015-03-22)
----------------

- Initial release.


tests/conftest.py
import multiprocessing
import os
import pytest
import shutil
import signal
import sys
import tempfile


class AbortCondition(object):

    def __init__(self, text='Running'):
        self.text = text

    def check_err(self, out, err):
        return self.text in err


def outerr_append(out, err, src):
    # append contents read by capsys.readouterr/capfd.readouterr to
    # out, err
    new_out, new_err = src.readouterr()
    if new_out:
        out += new_out
    if new_err:
        err += new_err
    return out, err


class SubprocessRunner(object):
    """Runs a passed in Python function (with args, kws).

    Desigend for use with server processes that must be aborted. After
    `run()` was called, we execute the passed in command, wait until
    we meet the abort condition evaluated by `abort_when(out, err)`
    and return the output.

    `abort_when` looks by default for string `Running` in stderr.
    """

    #: The condition to meet to abort a started server process
    abort_when = AbortCondition().check_err

    #: The exit code of the last run started (and stopped) process
    exit_code = None

    #: The sys.argv set during run()
    argv = sys.argv

    def __init__(self, capfd):
        self.capfd = capfd

    def run(self, target, *args, **kw):
        old_argv = sys.argv
        sys.argv = self.argv
        p1 = multiprocessing.Process(target=target, args=args, kwargs=kw)
        p1.start()
        timeout = 0.1
        out, err = ('', '')
        while timeout <= 103.0:  # abort after about 10 rounds
            p1.join(timeout)
            timeout *= 2
            out, err = outerr_append(out, err, self.capfd)
            if self.abort_when(out, err):
                break
        if p1.is_alive():
            # do not use p1.terminate() here, as only with SIGINT we get
            # coverage data from subprocess (terminate() sends SIGTERM).
            os.kill(p1.pid, signal.SIGINT)
            p1.join()
        self.exitcode = p1.exitcode
        out, err = outerr_append(out, err, self.capfd)
        sys.argv = old_argv
        return out, err


@pytest.fixture(scope="function")
def proc_runner(request, capfd):
    """Returns a SubprocessRunner capturing output via `capfd`.
    """
    runner = SubprocessRunner(capfd)
    return runner


@pytest.fixture(scope="session", autouse=True)
def temporary_tempdir_env(request):
    """Set a new root for temporary files.

    All temporary files created in a test session by `tempfile` module
    functions, are created in a folder created here.

    This folder is removed after test session.
    """
    old_tempdir = tempfile.gettempdir()
    new_tempdir = tempfile.mkdtemp()
    tempfile.tempdir = new_tempdir

    def restore_old_tempdir():
        tempfile.tempdir = old_tempdir
        shutil.rmtree(new_tempdir)

    request.addfinalizer(restore_old_tempdir)


tests/test_dropafile.py
# tests for dropafile module.
import base64
import math
import os
import pytest
import re
import shutil
import subprocess
import tempfile
from contextlib import contextmanager
from io import BytesIO
from werkzeug.datastructures import Headers
from werkzeug.test import Client, create_environ, EnvironBuilder
from werkzeug.wrappers import BaseResponse, Request
from dropafile import (
    DropAFileApplication, execute_cmd, create_ssl_cert, get_random_password,
    ALLOWED_PWD_CHARS, handle_options, run_server, get_store_path
    )


#: find a certificate path in output.
RE_CERTPATH = re.compile(
    '^.*Certificate in:[\s]+([^\s][^\n]+)\n.*$', re.M + re.S)


@contextmanager
def popen(*args, **kw):
    # a Python 2.6/2.7 compatible Popen contextmanager
    p = subprocess.Popen(*args, **kw)
    try:
        yield p
    finally:
        for stream in (p.stdout, p.stderr, p.stdin):
            if stream:
                stream.close()
        p.wait()


def encode_creds(username='somename', password=''):
    # turn credentials given into base64 encoded string
    auth_string = '%s:%s' % (username, password)
    encoded = base64.b64encode(auth_string.encode('utf-8'))
    return 'Basic %s' % encoded.decode('utf-8')


def get_basic_auth_headers(username='somename', password=''):
    # get a set of request headers with authorization set to creds.
    headers = Headers()
    headers.add(
        'Authorization', encode_creds(username=username, password=password))
    return headers


class TestHelpers(object):

    @pytest.mark.skipif(
        not os.path.exists('/bin/echo'), reason="needs /bin/echo")
    def test_excute_cmd(self):
        # we can execute commands (w/o shell)
        cmd = ["/bin/echo", "Hello $PATH"]
        out, err = execute_cmd(cmd)
        assert out == b'Hello $PATH\n'
        assert err == b''

    def test_create_cert(self, tmpdir):
        # we can create x509 certs
        path = tmpdir.dirname
        cert_path, key_path = create_ssl_cert(path)
        assert os.path.isfile(cert_path)
        assert os.path.isfile(key_path)
        shutil.rmtree(os.path.dirname(cert_path))

    def test_create_cert_no_path(self):
        # w/o a given path, one will be created for us
        cert_path, key_path = create_ssl_cert()
        assert os.path.isfile(cert_path)
        shutil.rmtree(os.path.dirname(cert_path))

    def test_get_random_password(self):
        # we can get a random password
        allowed_chars = '[A-HJ-NP-Z2-9a-hjkmnp-z]'
        RE_PWD = re.compile('^%s+$' % allowed_chars)
        password = get_random_password()
        assert RE_PWD.match(password)

    def test_get_random_password_entropy(self):
        # the entropy delivered by default >= 128 bits
        unique_chars = ''.join(list(set(ALLOWED_PWD_CHARS)))
        entropy_per_char = math.log(len(unique_chars)) / math.log(2)
        password = get_random_password()
        assert len(password) * entropy_per_char >= 128

    def test_get_store_path(self):
        # we can get a safe storage path
        store_dir = tempfile.mkdtemp()
        path = get_store_path(store_dir, 'test.txt')
        assert path == os.path.join(store_dir, 'test.txt')

    def test_get_store_path_one_file_in(self):
        # with one file in we get a modified filename
        store_dir = tempfile.mkdtemp()
        open(os.path.join(store_dir, 'test.txt'), 'w').write('foo')
        path = get_store_path(store_dir, 'test.txt')
        assert path.endswith('/test.txt-1')
        open(path, 'w').write('foo')
        path = get_store_path(store_dir, 'test.txt')
        assert path.endswith('/test.txt-2')
        open(path, 'w').write('foo')
        path = get_store_path(store_dir, 'test.txt')
        assert path.endswith('/test.txt-3')

    def test_get_store_path_two_files_in(self):
        # with two files in we also get a modified filename
        store_dir = tempfile.mkdtemp()
        open(os.path.join(store_dir, 'test.txt'), 'w').write('foo')
        open(os.path.join(store_dir, 'test.txt-2'), 'w').write('foo')
        path = get_store_path(store_dir, 'test.txt')
        assert path.endswith('/test.txt-1')


class TestApp(object):
    # no browser tests here

    def test_app_has_password(self):
        # DropAFileApplications have a password
        app = DropAFileApplication()
        assert hasattr(app, 'password')
        assert len(app.password) > 5

    def test_app_accepts_passwod(self):
        # DropAFileApps accept passwords passed in
        app = DropAFileApplication(password='verysecret')
        assert app.password == 'verysecret'

    def test_check_auth_requires_auth(self):
        # we require at least some creds to authenticate
        app = DropAFileApplication()
        app.password = 'sosecret'
        env = create_environ()
        request = Request(env)
        assert app.check_auth(request) is False

    def test_check_auth_wrong_passwd(self):
        # of course check_auth requires the correct password
        app = DropAFileApplication()
        app.password = 'sosecret'
        env = create_environ()
        env.update(HTTP_AUTHORIZATION=encode_creds(
            username='somename', password='wrong-password'))
        request = Request(env)
        assert app.check_auth(request) is False

    def test_check_auth_correct_passwd(self):
        # the correct password can be seen.
        app = DropAFileApplication()
        app.password = 'sosecret'
        env = create_environ()
        env.update(HTTP_AUTHORIZATION=encode_creds(
            username='somename', password='sosecret'))
        request = Request(env)
        assert app.check_auth(request) is True

    def test_handle_uploaded_files(self):
        # we can send files (that are stored)
        app = DropAFileApplication()
        builder = EnvironBuilder(
            method='POST',
            data={'file': (BytesIO(b'foo'), 'test.txt')}
            )
        req = Request(builder.get_environ())
        app.handle_uploaded_files(req)
        expected_path = os.path.join(app.upload_dir, 'test.txt')
        assert os.path.isfile(expected_path)
        assert open(expected_path, 'r').read() == 'foo'

    def test_handle_uploaded_files_wrong_formfield_name(self):
        # only files with form-name 'file' are considered
        app = DropAFileApplication()
        builder = EnvironBuilder(
            method='POST',
            data={'not_file': (BytesIO(b'foo'), 'test.txt')}
            )
        req = Request(builder.get_environ())
        app.handle_uploaded_files(req)
        assert os.listdir(app.upload_dir) == []

    def test_handle_uploaded_files_multiple_at_once(self):
        # we only take one file, even if multiple are offered
        app = DropAFileApplication()
        builder = EnvironBuilder(
            method='POST',
            data={'file': (BytesIO(b'foo'), 'test.txt'),
                  'file2': (BytesIO(b'bar'), 'test2.txt')}
            )
        req = Request(builder.get_environ())
        app.handle_uploaded_files(req)
        assert os.listdir(app.upload_dir) == ['test.txt']

    def test_handle_uploaded_files_output(self, capsys):
        # sent files are listed on commandline
        app = DropAFileApplication()
        builder = EnvironBuilder(
            method='POST',
            data={'file': (BytesIO(b'foo'), 'test.txt')}
            )
        req = Request(builder.get_environ())
        app.handle_uploaded_files(req)
        out, err = capsys.readouterr()
        assert 'RECEIVED:' in out
        assert 'test.txt' in out

    def test_handle_uploaded_files_no_files(self, capsys):
        # we notice if no files was sent (and do nothing)
        app = DropAFileApplication()
        req = Request(create_environ())
        app.handle_uploaded_files(req)
        out, err = capsys.readouterr()
        assert os.listdir(app.upload_dir) == []
        assert 'RECEIVED' not in out

    def test_handle_uploaded_files_not_overwritten(self):
        # we do not overwrite uploaded files
        app = DropAFileApplication()
        builder = EnvironBuilder(
            method='POST',
            data={'file': (BytesIO(b'uploaded'), 'test.txt')}
            )
        req = Request(builder.get_environ())
        upload_path = os.path.join(app.upload_dir, 'test.txt')
        with open(upload_path, 'w') as fd:
            fd.write('original')
        app.handle_uploaded_files(req)
        assert sorted(
            os.listdir(app.upload_dir)) == ['test.txt', 'test.txt-1']
        with open(os.path.join(app.upload_dir, 'test.txt'), 'r') as fd:
            content1 = fd.read()
        with open(os.path.join(app.upload_dir, 'test.txt-1'), 'r') as fd:
            content2 = fd.read()
        assert content1 == 'original'
        assert content2 == 'uploaded'


class TestArgParser(object):

    def test_help(self, capsys):
        # we support --help
        with pytest.raises(SystemExit) as exc_info:
            handle_options(['foo', '--help'])
        out, err = capsys.readouterr()
        assert exc_info.value.code == 0

    def test_help_lists_all_options(self, capsys):
        # all options are listed in --help
        with pytest.raises(SystemExit):
            handle_options(['foo', '--help'])
        out, err = capsys.readouterr()
        assert '--host' in out
        assert '--port' in out
        assert '--secret' in out

    def test_defaults(self):
        # we can get options with defaults set
        result = handle_options([])
        assert result is not None
        assert result.host == 'localhost'
        assert result.port == 8443
        assert result.secret is None

    def test_host(self):
        result = handle_options(['--host', 'foo'])
        assert result.host == 'foo'

    def test_port(self):
        result = handle_options(['--port', '1234'])
        assert result.port == 1234

    def test_secret(self):
        result = handle_options(['--secret', 'sosecret'])
        assert result.secret == 'sosecret'


class Test_run_server(object):

    def test_no_options(self, proc_runner):
        # we can start a server (no options given)
        proc_runner.argv = ['dropafile', ]
        out, err = proc_runner.run(run_server, args=None)
        assert 'Certificate in:' in out
        assert 'Running' in err

    def test_help(self, proc_runner):
        # we can get help from run_server()
        out, err = proc_runner.run(run_server, args=["dropafile", "--help"])
        assert 'show this help message and exit' in out
        assert proc_runner.exitcode == 0

    def test_secret(self, proc_runner):
        # a passed-in password is respected
        out, err = proc_runner.run(
            run_server, args=["dropafile", "-s", "sosecret"])
        assert 'Password is: sosecret' in out

    def test_host_and_port(self, proc_runner):
        # we can set a custom host and port we want to bind to
        out, err = proc_runner.run(
            run_server, args=["dropafile",
                              "--host", "0.0.0.0",
                              "--port", "12345"]
            )
        assert "https://0.0.0.0:12345/" in err


class TestFunctional(object):
    # Functional browser tests

    def test_page_response(self):
        # we can get some HTML page for any path
        application = DropAFileApplication()
        client = Client(application, BaseResponse)
        headers = get_basic_auth_headers(
            username='somename', password=application.password)
        resp = client.get('/', headers=headers)
        assert resp.status == '200 OK'
        mimetype = resp.headers.get('Content-Type')
        assert mimetype == 'text/html; charset=utf-8'

    def test_get_js(self):
        # we can get the dropzonejs JavaScript
        application = DropAFileApplication()
        client = Client(application, BaseResponse)
        headers = get_basic_auth_headers(
            username='somename', password=application.password)
        resp = client.get('dropzone.js', headers=headers)
        assert resp.status == '200 OK'
        mimetype = resp.headers.get('Content-Type')
        assert mimetype == 'text/javascript; charset=utf-8'

    def test_get_css(self):
        # we can get the dropzonejs CSS
        application = DropAFileApplication()
        client = Client(application, BaseResponse)
        headers = get_basic_auth_headers(
            username='somename', password=application.password)
        resp = client.get('dropzone.css', headers=headers)
        assert resp.status == '200 OK'
        mimetype = resp.headers.get('Content-Type')
        assert mimetype == 'text/css; charset=utf-8'

    def test_send_file(self):
        # we can send files
        application = DropAFileApplication()
        client = Client(application, BaseResponse)
        headers = get_basic_auth_headers(
            username='somename', password=application.password)
        resp = client.post(
            '/index.html',
            headers=headers,
            data={
                'file': (BytesIO(b'Some Content'), 'sample.txt'),
                },
            )
        assert resp.status == '200 OK'
        uploaded_path = os.path.join(application.upload_dir, 'sample.txt')
        assert os.path.isfile(uploaded_path)

    def test_unauthorized_by_default(self):
        # By default we get an Unauthorized message
        app = DropAFileApplication()
        client = Client(app, BaseResponse)
        resp = client.get('/')
        assert resp.status == '401 UNAUTHORIZED'

    def test_basic_auth_req_by_default(self):
        # By default we require basic auth from client
        app = DropAFileApplication()
        client = Client(app, BaseResponse)
        resp = client.get('/')
        header = resp.headers.get('WWW-Authenticate', None)
        assert header is not None

    def test_page_set_password(self):
        # we can get some HTML page for any path
        application = DropAFileApplication(password="sosecret")
        client = Client(application, BaseResponse)
        headers = get_basic_auth_headers(
            username='somename', password="sosecret")
        resp = client.get('/', headers=headers)
        assert resp.status == '200 OK'


dropafile.py
#    dropafile -- drop me a file on a webpage
#    Copyright (C) 2015  Uli Fouquet
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""dropafile - Drop a file on a webpage.
"""
import argparse
import os
import random
import pkg_resources
import ssl
import subprocess
import sys
import tempfile
from werkzeug import secure_filename
from werkzeug.serving import run_simple
from werkzeug.wrappers import Request, Response


#: Official version
__version__ = pkg_resources.get_distribution('dropafile').version


PATH_MAP = {
    '/dropzone.js': ('dropzone.js', 'text/javascript'),
    '/dropzone.css': ('dropzone.css', 'text/css'),
    '/style.css': ('style.css', 'text/css'),
    '/index.html': ('page.html', 'text/html'),
    }


STATIC_DIR = os.path.join(os.path.dirname(__file__), 'static')


#: Chars allowed in passwords.
#: We allow plain ASCII chars and numbers, with some entitites removed,
#: that can be easily mixed up: letter `l` and number one, for instance.
ALLOWED_PWD_CHARS = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789abcdefghjkmnpqrstuvwxyz'


def handle_options(args):
    """Handle commandline options.

    Expects the arguments passed to dropafile as a list of
    arguments. `args` is expected to represent ``sys.argv[1:]``,
    i.e. the arguments when called, without the programme name.

    Returns parsed options as provided by :mod:`argparse`.
    """
    parser = argparse.ArgumentParser(description="Start dropafile app.")
    parser.add_argument(
        '--host', required=False, default='localhost',
        help=(
            'Host we bind to. An IP address or DNS name. `localhost`'
            ' by default.'
            ),
        )
    parser.add_argument(
        '-p', '--port', required=False, default=8443, type=int,
        help=(
            'Port we listen at. An integer. 8443 by default.'
            )
        )
    parser.add_argument(
        '-s', '--secret', required=False, metavar='PASSWORD',
        help=(
            'Password to access dropafile. If none is given we generate '
            'one.'
            )
        )
    opts = parser.parse_args(args)
    return opts


def get_random_password():
    """Get a password generated from `ALLOWED_PWD_CHARS`.

    The password entropy should be >= 128 bits. We use `SystemRandom()`,
    which should provide enough randomness to work properly.
    """
    rnd = random.SystemRandom()
    return ''.join(
        [rnd.choice(ALLOWED_PWD_CHARS) for x in range(23)])


def get_store_path(directory, filename):
    """Get a path where we can safely store a file.

    The file should be stored in `directory` under name `filename`.
    If `filename` already exists in `directory`, we construct new
    names by appending '-<NUM>' to the original filename, where
    ``<NUM>`` is a number counting up.
    """
    filename = secure_filename(filename)
    path = os.path.join(directory, filename)
    num = 1
    while os.path.exists(path):
        path = os.path.join(directory, '%s-%s' % (filename, num))
        num += 1
    return path


class DropAFileApplication(object):
    """Drop-A-File application.

    A `werkzeug` based WSGI application providing a basic-auth
    protected web interface for file uploads.

    `password` is required to access the application's service. If
    none is provided, we generate one for you.

    `upload_dir` is the directory, where we store files uploaded by
    users. If none is given we create a temporary directory on
    start-up. Please note: the directory is not removed on shutdown.
    """

    #: the password we require (no username neccessary)
    password = None

    #: a path where we store files uploaded by users.
    upload_dir = None

    def __init__(self, password=None, upload_dir=None):
        if password is None:
            password = get_random_password()
        self.password = password
        if upload_dir is None:
            upload_dir = tempfile.mkdtemp()
        self.upload_dir = upload_dir

    def check_auth(self, request):
        """Check basic auth against local password.

        We accept any username, but only *the* one password. Returns
        ``True`` in case of success, ``False`` otherwise.

        `request` must contain basic-auth authorization headers (as
        set by browsers) to succeed.
        """
        auth = request.authorization
        if auth is None:
            return False
        if auth.password != self.password:
            return False
        return True

    def authenticate(self):
        """Send 401 requesting basic auth from client.

        Send back 401 response to client. Contains some HTML to
        display an 'Unauhorized' page. Should make browsers ask users
        for username and password.
        """
        return Response(
            '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n'
            '<title>401 Unauthorized</title>\n'
            '<h1>Unauthorized</h1>'
            '<p>You are not authorized to use this service.</p>',
            401, {'WWW-Authenticate': 'Basic realm="Login required"',
                  'Content-Type': 'text/html'}
            )

    def handle_uploaded_files(self, request):
        """Look for an upload file in `request`.

        If one is found, it is saved to `self.upload_dir`.
        """
        uploaded_file = request.files.get('file', None)
        if uploaded_file is None:
            return
        filename = secure_filename(uploaded_file.filename)
        path = os.path.join(self.upload_dir, filename)
        print("RECEIVED: %s" % path)
        uploaded_file.save(path)

    @Request.application
    def __call__(self, request):
        if not self.check_auth(request):
            return self.authenticate()
        self.handle_uploaded_files(request)
        path = request.path
        if path not in PATH_MAP.keys():
            path = '/index.html'
        filename, mimetype = PATH_MAP[path]
        with open(os.path.join(STATIC_DIR, filename)) as fd:
            page = fd.read()
        return Response(page, mimetype=mimetype)


def execute_cmd(cmd_list):
    """Excute the command `cmd_list`.

    Returns stdout and stderr output.
    """
    pipe = subprocess.PIPE
    proc = subprocess.Popen(
        cmd_list, stdout=pipe, stderr=pipe, shell=False)
    try:
        stdout, stderr = proc.communicate()
    finally:
        proc.stdout.close()
        proc.stderr.close()
        proc.wait()
    return stdout, stderr


def create_ssl_cert(path=None, bits=4096, days=2, cn='localhost',
                    country='US', state='', location=''):
    """Create an SSL RSA cert and key in directory `path`.

    Returns a tuple `(certificate_path, key_path)`.

    `path`
      A directory, where certificate and key can be stored. If none is
      given, we create a temporary one.

    Default attribute values of the certificate are read from a
    package-local SSL configuration file ``openssl.conf``.


    Some attribute values can be overridden:

    `bits`
      number of bits of the key.

    `days`
      number of days of validity of the generated certificate.

    `cn`
      Common Name. Put the domain under which the app will be
      served in here.

    `state` and `location`
      will be empty by default.
    """
    print("Creating temporary self-signed SSL certificate...")
    if path is None:
        path = tempfile.mkdtemp()
    cert_path = os.path.join(path, 'cert.pem')
    key_path = os.path.join(path, 'cert.key')
    openssl_conf = os.path.join(os.path.dirname(__file__), 'openssl.conf')
    subject = '/C=%s/ST=%s/L=%s/O=%s/OU=%s/CN=%s/emailAddress=%s/' % (
        country, state, location, '', '', cn, '')
    cmd = [
        'openssl', 'req', '-x509', '-newkey', 'rsa:%s' % bits, '-nodes',
        '-out', cert_path, '-keyout', key_path, '-days', '%s' % days,
        '-sha256', '-config', openssl_conf, '-batch', "-subj", subject
        ]
    out, err = execute_cmd(cmd)
    print("Done.")
    print("Certificate in: %s" % cert_path)
    print("Key in:         %s" % key_path)
    return cert_path, key_path


def get_ssl_context(cert_path=None, key_path=None):
    """Get an SSL context to serve HTTP.

    If `cert_path` or `key_path` are ``None``, we create some. Then we
    add some modifiers (avail. with Python >= 2.7.9) to disable unsafe
    ciphers etc.

    The returned SSL context can be used with Werkzeug `run_simple`.
    """
    if (key_path is None) or (cert_path is None):
        cert_path, key_path = create_ssl_cert()
    ssl_context = (cert_path, key_path)
    if hasattr(ssl, 'SSLContext'):  # py >= 2.7.9
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
        ssl_context.options |= ssl.OP_NO_SSLv2  # considered unsafe
        ssl_context.options |= ssl.OP_NO_SSLv3  # considered unsafe
        ssl_context.load_cert_chain(cert_path, key_path)
    return ssl_context


def run_server(args=None):
    """Run a `werkzeug` server, serving a :class:`DropAFileApplication`.

    Called when running `dropafile` from commandline. Serves a
    :class:`DropAFileApplication` instance until aborted.

    Options `argv` are taken from commandline if not specified.

    Generates a password and temporary SSL certificate/key on startup
    unless otherwise requested in options/args.
    """
    if args is None:
        args = sys.argv
    options = handle_options(args[1:])
    ssl_context = get_ssl_context()
    sys.stdout.flush()
    application = DropAFileApplication(password=options.secret)
    print("Password is: %s" % application.password)
    sys.stdout.flush()
    run_simple(options.host, options.port, application,
               ssl_context=ssl_context)
